src/copy.c

```diff
@@ -74,6 +138,381 @@ copy_chunk_state_create(Hypertable *ht, Relation rel, CopyFromFunc from_func, Co
+			buffer = (TSCopyMultiInsertBuffer *) linitial(miinfo->multiInsertBuffers);
+		}
+
+		TSCopyMultiInsertBufferCleanup(miinfo, buffer);
+		miinfo->multiInsertBuffers = list_delete_first(miinfo->multiInsertBuffers);
+	}
+}
+
+/*
+ * Cleanup allocated buffers and free memory.
+ */
+static inline void
+TSCopyMultiInsertInfoCleanup(TSCopyMultiInsertInfo *miinfo)
+{
+	ListCell *lc;
+
+	foreach (lc, miinfo->multiInsertBuffers)
+		TSCopyMultiInsertBufferCleanup(miinfo, lfirst(lc));
+
+	list_free(miinfo->multiInsertBuffers);
+}
+
+/*
+ * Get the next TupleTableSlot that the next tuple should be stored in.
+ *
+ * Callers must ensure that the buffer is not full.
+ *
+ * Note: 'miinfo' is unused but has been included for consistency with the
+ * other functions in this area.
+ */
+static inline TupleTableSlot *
+TSCopyMultiInsertInfoNextFreeSlot(TSCopyMultiInsertInfo *miinfo, ChunkInsertState *cis)
+{
+	TSCopyMultiInsertBuffer *buffer = cis->copy_multi_insert_buffer;
+	int nused = buffer->nused;
+
+	Assert(buffer != NULL);
+	Assert(nused < MAX_BUFFERED_TUPLES);
+
+	if (buffer->slots[nused] == NULL)
+		buffer->slots[nused] = table_slot_create(cis->result_relation_info->ri_RelationDesc, NULL);
+	return buffer->slots[nused];
+}
+
+/*
+ * Record the previously reserved TupleTableSlot that was reserved by
+ * TSCopyMultiInsertInfoNextFreeSlot as being consumed.
+ */
+static inline void
+TSCopyMultiInsertInfoStore(TSCopyMultiInsertInfo *miinfo, ResultRelInfo *rri, ChunkInsertState *cis,
+						   TupleTableSlot *slot, CopyFromState cstate)
+{
+	TSCopyMultiInsertBuffer *buffer = cis->copy_multi_insert_buffer;
+
+	Assert(buffer != NULL);
+	Assert(slot == buffer->slots[buffer->nused]);
+
+	/* Store the line number so we can properly report any errors later */
+	uint64 lineno = 0;
+
+	/* The structure CopyFromState is private in PG < 14. So we can not access
+	 * the members like the line number or the size of the tuple.
+	 */
+#if PG14_GE
+	if (cstate != NULL)
+		lineno = cstate->cur_lineno;
+#endif
+	buffer->linenos[buffer->nused] = lineno;
+
+	/* Record this slot as being used */
+	buffer->nused++;
+
+	/* Update how many tuples are stored and their size */
+	miinfo->bufferedTuples++;
+
+#if PG14_GE
+	if (cstate != NULL)
+	{
+		int tuplen = cstate->line_buf.len;
+		miinfo->bufferedBytes += tuplen;
+	}
+	else
+	{
+#else
+	/*
+	 * Determine the size of the tuple by calculating the size of the per-tuple
+	 * memory context. The per-tuple memory context is deleted in the copyfrom
+	 * function per processed tuple. So, at this time only the tuple is stored
+	 * in this context.
+	 */
+	MemoryContextCounters context_counter;
+	MemoryContext tuple_context = GetPerTupleMemoryContext(miinfo->estate);
+	tuple_context->methods->stats(tuple_context, false, 0, &context_counter);
+	Size used_memory = context_counter.totalspace - context_counter.freespace;
+	Assert(used_memory > 0);
+	miinfo->bufferedBytes += used_memory;
```

Comment:
I'm confused by the ifdefs, does this `else` block end up empty for pg >= 14?

TBH I'd just copy the private struct definitions here. They are not so likely to change in minor releases.

Reply:
When the code runs on >= PG14 and the `cstate` is present, the memory consumption is calculated based on this. Otherwise, the memory context is used to calculate the size.

It's a good hint that the structures don't change that often. Then I will copy them next time.

Reply:
I mean, suppose PG14_GE is true, then it becomes:
```
     if (cstate != NULL)
     {
          int tuplen = cstate->line_buf.len;
          miinfo->bufferedBytes += tuplen;
     }
     else
     {
     }
```

Looks like this empty `else` block is not needed?

Reply:
Maybe you meant "#endif" instead of "#else" in line 497?

Reply:
Fixed